mono/metadata/threads.c

```diff
@@ -6583,6 +6582,353 @@ summarizer_supervisor_end (SummarizerSupervisorState *state)
+	MonoSemType begin_crash_report;
+	/* originator cant post commands to the leader */
+	MonoSemType leader_commanded;
+	int leader_command;
+	/* pipe to communicate back from the summarizer leader to the originator */
+	int response_fds[2];
+	/* Only one orignator at a time, gated by mono_threads_summarize tickets */
+	MonoSummarizerOriginator originator;
+} MonoSummarizerLeader;
+
+static MonoSummarizerLeader summarizer_leader_data;
+
+/* Leader state machine */
+enum LeaderState {
+	LEADER_STATE_READY = 0,
+	LEADER_STATE_COLLECTING_IDS,
+	LEADER_STATE_WAIT_BEFORE_SUSPEND_OTHERS,
+	LEADER_STATE_SUSPEND_OTHERS,
+	LEADER_STATE_WAIT_FOR_ORIGINATOR_STACK,
+	LEADER_STATE_SUMMARIZER_STATE_TERM,
+};
+
+/* Commands from crash originator to the crash leader */
+enum LeaderCommand {
+	LEADER_COMMAND_ZERO = 0, /* not used */
+	LEADER_COMMAND_CANCEL = -1,
+	LEADER_COMMAND_PROCEED_TO_SUSPEND = 1,
+	LEADER_COMMAND_PROCEED_TO_TERM = 2,
+};
+
+/* Responses from the crash leader to the crash originator */
+enum LeaderResponse {
+	LEADER_RESPONSE_IDS_COLLECTED = 1,
+	LEADER_RESPONSE_THREADS_SUSPENDED = 2,
+	LEADER_RESPONSE_STACKS_WALKED = 3,
+};
+
+#undef LEADER_DEBUG
+
+#ifdef LEADER_DEBUG
+#define LEADER_LOG(...) g_async_safe_printf (__VA_ARGS__)
+#else
+#define LEADER_LOG(...) /*empty*/
+#endif
+
+/* Called by the leader to send responses to the originator */
+static void
+summarizer_leader_response_write (char b)
+{
+	int res;
+	LEADER_LOG ("leader --> originator: %d\n", (int)b);
+	while ((res = write (summarizer_leader_data.response_fds[1], &b, sizeof (b))) < 0 && errno == EINTR);
+}
+
+/* Called by the originator to receive (blocking) responses from the leader */
+static int
+summarizer_leader_response_read (void)
+{
+	char buf;
+	int nread = 0;
+	do {
+		int res = read(summarizer_leader_data.response_fds[0], &buf, sizeof (buf));
+		if (res < 0) {
+			if (errno == EINTR)
+				continue;
+			else
+				return -1;
+		}
+		nread += res;
+	} while (nread < sizeof (buf));
+	LEADER_LOG ("originator <---- leader : %d\n", (int)buf);
+	return (int)buf;
+}
+
+#define STATE_CASE(x) case x: return #x
+
+static const char *
+leader_state_name (int leader_state) {
+	switch (leader_state) {
+#define STATE_CASE(x) case x: return #x
+		STATE_CASE(LEADER_STATE_READY);
+		STATE_CASE(LEADER_STATE_COLLECTING_IDS);
+		STATE_CASE(LEADER_STATE_WAIT_BEFORE_SUSPEND_OTHERS);
+		STATE_CASE(LEADER_STATE_SUSPEND_OTHERS);
+		STATE_CASE(LEADER_STATE_WAIT_FOR_ORIGINATOR_STACK);
+		STATE_CASE(LEADER_STATE_SUMMARIZER_STATE_TERM);
+#undef STATE_CASE
+	default:
+		return "<unknown leader state>";
+	}
+}
+
+/* Called by the leader to wait for a command from the crash originator */
+static gboolean
+summarizer_leader_wait_for_command (int *leader_command, int *leader_state)
+{
+	MONO_ENTER_GC_SAFE;
+	/* allow interruptions */
+	while (mono_os_sem_wait (&summarizer_leader_data.leader_commanded, MONO_SEM_FLAGS_ALERTABLE) < 0);
+	MONO_EXIT_GC_SAFE;
+	mono_memory_barrier ();
```

Comment:
This barrier is redundant. Synchronization operations have implicit memory barriers, otherwise they wouldn't really be synchronizing :)